File: AdisyonUygulamaSwiftAutoLayout/AdisyonUygulamaSwiftAutoLayout/services/AdisyonService.swift
//
//  AdisyonService.swift
//  AdisyonUygulamaSwiftAutoLayout
//
//  Created by Alperen Saraç on 6.06.2025.
//

import Foundation

// MARK: - Response Models (Placeholder)
struct UrunSilResponse: Codable {
    // örn: let success: Bool
}

struct KategoriSilResponse: Codable {
    // örn: let success: Bool
}

// MARK: - Service

final class AdisyonService {
    static let shared = AdisyonService()
    private init() {}

    private let baseURL = URL(string: "https://alperensaracdeneme.com/adisyon/")!  // “https://your.api.base.url/” kısmını kendi sunucu adresinizle değiştirin.

    private let jsonDecoder: JSONDecoder = {
        let decoder = JSONDecoder()
        // Eğer tarih formatları veya farklı anahtar stilleri kullanılıyorsa buraya ekleyin.
        return decoder
    }()

    // Helper: Form URL-encoded body oluşturma
    private func makeFormBody(from parameters: [String: Any]) -> Data? {
        let queryItems = parameters.map { key, value in
            URLQueryItem(name: key, value: "\(value)")
        }
        var components = URLComponents()
        components.queryItems = queryItems
        return components.percentEncodedQuery?.data(using: .utf8)
    }

    // Helper: Genel POST isteği
    private func postRequest(
        path: String,
        formParameters: [String: Any]
    ) async throws -> Data {
        let url = baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = makeFormBody(from: formParameters)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResp = response as? HTTPURLResponse,
              200..<300 ~= httpResp.statusCode else {
            throw URLError(.badServerResponse)
        }
        return data
    }

    // Helper: Genel GET isteği (sorgu parametreli veya parametresiz)
    private func getRequest(
        path: String,
        queryParameters: [String: String]? = nil
    ) async throws -> Data {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if let queries = queryParameters {
            components.queryItems = queries.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        guard let url = components.url else {
            throw URLError(.badURL)
        }

        let (data, response) = try await URLSession.shared.data(from: url)
        guard let httpResp = response as? HTTPURLResponse,
              200..<300 ~= httpResp.statusCode else {
            throw URLError(.badServerResponse)
        }
        return data
    }

    // MARK: - API Methods

    /// GET "masa_listesi.php" → [Masa]
    func getMasalar() async throws -> [Masa] {
        let data = try await getRequest(path: "masa_listesi.php")
        return try jsonDecoder.decode([Masa].self, from: data)
    }

    /// GET "masa_urunleri.php?masa_id={masaId}" → [MasaUrun]
    func getMasaUrunleri(masaId: Int) async throws -> [MasaUrun] {
        let params = ["masa_id": "\(masaId)"]
        let data = try await getRequest(path: "masa_urunleri.php", queryParameters: params)
        return try jsonDecoder.decode([MasaUrun].self, from: data)
    }

    /// POST "masa_sil.php" (FormUrlEncoded) { "masa_id": masaId } → ResponseBody (ham veri)
    func masaSil(masaId: Int) async throws -> Data {
        return try await postRequest(
            path: "masa_sil.php",
            formParameters: ["masa_id": masaId]
        )
    }

    /// POST "urun_ekle.php" (FormUrlEncoded)
    /// { "urun_ad": urunAd, "urun_fiyat": urunFiyat, "urun_kategori": urunKategori, "urun_adet": urunAdet, "urun_resim": urunResimBase64 }
    func urunEkle(
        urunAd: String,
        urunFiyat: Float,
        urunKategori: Int,
        urunAdet: Int,
        urunResimBase64: String
    ) async throws -> Data {
        let params: [String: Any] = [
            "urun_ad": urunAd,
            "urun_fiyat": urunFiyat,
            "urun_kategori": urunKategori,
            "urun_adet": urunAdet,
            "urun_resim": urunResimBase64
        ]
        return try await postRequest(path: "urun_ekle.php", formParameters: params)
    }

    /// POST "urun_sil.php" (FormUrlEncoded) { "urun_ad": urunAd } → UrunSilResponse
    func urunSil(urunAd: String) async throws -> UrunSilResponse {
        let data = try await postRequest(
            path: "urun_sil.php",
            formParameters: ["urun_ad": urunAd]
        )
        return try jsonDecoder.decode(UrunSilResponse.self, from: data)
    }

    /// GET "masa_urun_ekle.php?masa_id={masaId}&urun_id={urunId}&adet={adet}" → ResponseBody (ham veri)
    func urunEkle(masaId: Int, urunId: Int, adet: Int) async throws -> Data {
        let params = [
            "masa_id": "\(masaId)",
            "urun_id": "\(urunId)",
            "adet": "\(adet)"
        ]
        let data = try await getRequest(path: "masa_urun_ekle.php", queryParameters: params)
        return data
    }

    /// POST "masa_birlestir.php" (FormUrlEncoded)
    /// { "ana_masa_id": anaMasaId, "birlestirilecek_masa_id": birlestirilecekMasaId }
    func masaBirlestir(anaMasaId: Int, birlestirilecekMasaId: Int) async throws -> Data {
        let params: [String: Any] = [
            "ana_masa_id": anaMasaId,
            "birlestirilecek_masa_id": birlestirilecekMasaId
        ]
        return try await postRequest(path: "masa_birlestir.php", formParameters: params)
    }

    /// POST "masa_ekle.php" (FormUrlEncoded) — parametre yok
    func masaEkle() async throws -> Data {
        return try await postRequest(path: "masa_ekle.php", formParameters: [:])
    }

    /// GET "urunleri_getir.php" → [Urun]
    func getUrunler() async throws -> [Urun] {
        let data = try await getRequest(path: "urunleri_getir.php")
        return try jsonDecoder.decode([Urun].self, from: data)
    }

    /// POST "masa_odeme.php" (FormUrlEncoded) { "masa_id": masaId } → ResponseBody (ham veri)
    func masaOde(masaId: Int) async throws -> Data {
        return try await postRequest(path: "masa_odeme.php", formParameters: ["masa_id": masaId])
    }

    /// GET "masa_toplam_fiyat.php?masa_id={masaId}" → [String: Float]
    func getToplamFiyat(masaId: Int) async throws -> [String: Float] {
        let params = ["masa_id": "\(masaId)"]
        let data = try await getRequest(path: "masa_toplam_fiyat.php", queryParameters: params)
        return try jsonDecoder.decode([String: Float].self, from: data)
    }

    /// POST "urun_cikar.php" (FormUrlEncoded) { "masa_id": masaId, "urun_id": urunId } → ResponseBody (ham veri)
    func urunCikar(masaId: Int, urunId: Int) async throws -> Data {
        let params: [String: Any] = [
            "masa_id": masaId,
            "urun_id": urunId
        ]
        return try await postRequest(path: "urun_cikar.php", formParameters: params)
    }

    /// GET "kategorileri_getir.php" → [Kategori]
    func getKategoriler() async throws -> [Kategori] {
        let data = try await getRequest(path: "kategorileri_getir.php")
        return try jsonDecoder.decode([Kategori].self, from: data)
    }

    /// GET "masa_getir.php?masa_id={masaId}" → Masa
    func masaGetir(masaId: Int) async throws -> Masa {
        let params = ["masa_id": "\(masaId)"]
        let data = try await getRequest(path: "masa_getir.php", queryParameters: params)
        return try jsonDecoder.decode(Masa.self, from: data)
    }

    /// POST "kategori_ekle.php" (FormUrlEncoded) { "kategori_ad": kategoriAd } → ResponseBody (ham veri)
    func kategoriEkle(kategoriAd: String) async throws -> Data {
        return try await postRequest(path: "kategori_ekle.php", formParameters: ["kategori_ad": kategoriAd])
    }

    /// POST "kategori_sil.php" (FormUrlEncoded) { "kategori_id": id } → KategoriSilResponse
    func kategoriSil(id: Int) async throws -> KategoriSilResponse {
        let data = try await postRequest(path: "kategori_sil.php", formParameters: ["kategori_id": id])
        return try jsonDecoder.decode(KategoriSilResponse.self, from: data)
    }
}

File: adisyon_flutter/lib/service/AdisyonService.dart
// lib/service/AdisyonService.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import '../model/Kategori.dart';
import '../model/Masa.dart';
import '../model/MasaUrun.dart';
import '../model/Urun.dart';

class AdisyonService {
  AdisyonService({this.baseUrl = 'https://alperensaracdeneme.com/adisyon/'});

  final String baseUrl;

  /* ─────────────────────── HTTP helpers ─────────────────────── */

  Future<dynamic> _get(String path, [Map<String, dynamic>? query]) async {
    final uri = Uri.parse(baseUrl + path).replace(queryParameters: query);
    print('▶️ GET $uri');

    final resp = await http.get(uri);

    if (resp.statusCode != 200) {
      throw HttpException('GET ${resp.statusCode}: ${resp.body}');
    }

    try {
      return resp.body.isEmpty ? null : jsonDecode(resp.body);
    } catch (e) {
      print('❌ JSON parse hatası: $e\n📦 Gelen body:\n${resp.body}');
      return null;
    }
  }

  Future<dynamic> _post(String path, Map<String, dynamic> body) async {
    final uri = Uri.parse(baseUrl + path);
    final resp = await http.post(
      uri,
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: body,
    );

    if (resp.statusCode != 200) {
      throw HttpException('POST ${resp.statusCode}: ${resp.body}');
    }

    try {
      return resp.body.isEmpty ? null : jsonDecode(resp.body);
    } catch (e) {
      print('❌ JSON parse hatası: $e\n📦 Gelen body:\n${resp.body}');
      return null;
    }
  }

  /* ─────────────────────── MASA ─────────────────────── */

  Future<List<Masa>> getMasalar() async {
    final data = await _get('masa_listesi.php');
    if (data is List) {
      return data.map((e) => Masa.fromJson(e)).toList();
    }
    print('⚠️ masa_listesi.php beklenmedik veri döndürdü: $data');
    return [];
  }

  Future<Masa> masaGetir(int id) async {
    final data = await _get('masa_getir.php', {'masa_id': id});
    if (data is Map<String, dynamic>) {
      return Masa.fromJson(data);
    }
    print('⚠️ masa_getir.php beklenmedik veri döndürdü: $data');
    throw StateError("Masa getirilemedi");
  }

  Future<void> masaEkle() => _post('masa_ekle.php', {});
  Future<void> masaSil(int id) => _post('masa_sil.php', {'masa_id': id});
  Future<void> masaOde(int id) => _post('masa_odeme.php', {'masa_id': id});

  Future<void> masaBirlestir(int ana, int diger) => _post(
    'masa_birlestir.php',
    {'ana_masa_id': ana, 'birlestirilecek_masa_id': diger},
  );

  Future<List<MasaUrun>> getMasaUrunleri(int id) async {
    final data = await _get('masa_urunleri.php', {'masa_id': id});
    if (data is List) {
      return data.map((e) => MasaUrun.fromJson(e)).toList();
    }
    print('⚠️ masa_urunleri.php beklenmedik veri döndürdü: $data');
    return [];
  }

  Future<void> masaUrunEkle(int masa, int urun, int adet) =>
      _get('masa_urun_ekle.php', {
        'masa_id': masa,
        'urun_id': urun,
        'adet': adet,
      });

  Future<void> urunCikar(int masa, int urun) =>
      _post('urun_cikar.php', {'masa_id': masa, 'urun_id': urun});

  Future<Map<String, double>> getToplamFiyat(int id) async {
    final data = await _get('masa_toplam_fiyat.php', {'masa_id': id});
    if (data is Map) {
      return data.map((k, v) => MapEntry(k.toString(), (v as num).toDouble()));
    }
    print('⚠️ masa_toplam_fiyat.php beklenmedik veri döndürdü: $data');
    return {};
  }

  /* ─────────────────────── URUN ─────────────────────── */

  Future<List<Urun>> getUrunler() async {
    final data = await _get('urunleri_getir.php');
    if (data is List) {
      return data.map((e) => Urun.fromJson(e)).toList();
    }
    print('⚠️ urunleri_getir.php beklenmedik veri döndürdü: $data');
    return [];
  }

  Future<void> urunEkle(
      String ad,
      double fiyat,
      int kategoriId,
      int adet,
      String resim64,
      ) =>
      _post('urun_ekle.php', {
        'urun_ad': ad,
        'urun_fiyat': fiyat,
        'urun_kategori': kategoriId,
        'urun_adet': adet,
        'urun_resim': resim64,
      });

  Future<Map<String, dynamic>> urunSil(String ad) async {
    final res = await _post('urun_sil.php', {'urun_ad': ad});
    if (res is Map<String, dynamic>) return res;
    throw StateError('urun_sil.php beklenmedik çıktı: $res');
  }

  /* ─────────────────────── KATEGORI ─────────────────────── */

  Future<List<Kategori>> getKategoriler() async {
    final data = await _get('kategorileri_getir.php');
    print("🔍 kategori response: $data");

    if (data is List) {
      return data.map((e) => Kategori.fromJson(e)).toList();
    } else {
      print('❌ kategori veri tipi: ${data.runtimeType}, içerik: $data');
      return [];
    }
  }

  Future<void> kategoriEkle(String ad) =>
      _post('kategori_ekle.php', {'kategori_ad': ad});

  Future<Map<String, dynamic>> kategoriSil(int id) async {
    final res = await _post('kategori_sil.php', {'kategori_id': id});
    if (res is Map<String, dynamic>) return res;
    throw StateError('kategori_sil.php beklenmedik çıktı: $res');
  }
}
